docs/Generators-Cluster.md

```diff
@@ -86,20 +86,8 @@ The cluster selector also supports set-based requirements, as used by [several c
 ### Deploying to the local cluster
 
 In Argo CD, the 'local cluster' is the cluster upon which Argo CD (and the ApplicationSet controller) is installed. This is to distinguish it from 'remote clusters', which are those that are added to Argo CD [declaratively](https://argoproj.github.io/argo-cd/operator-manual/declarative-setup/#clusters) or via the [Argo CD CLI](https://argoproj.github.io/argo-cd/getting_started/#5-register-a-cluster-to-deploy-apps-to-optional).
- 
-The cluster generator will automatically target both local and non-local clusters, for every cluster that matches the cluster selector.
 
-If you wish to target only remote clusters with your Applications (e.g. you want to exclude the local cluster), then use a cluster selector with labels, for example:
-```yaml
-spec:
-  generators:
-  - clusters:
-      selector:
-        matchLabels:
-          argocd.argoproj.io/secret-type: cluster
-```
-
-This selector will not match the default local cluster, since the default local cluster does not have a Secret (and thus does not have the `argocd.argoproj.io/secret-type` label on that secret). Any cluster selector that selects on that label will automatically exclude the default local cluster.
+Where the selector is empty, the Cluster generator will automatically target all clusters, both local and non-local. Otherwise, it will discover only clusters with secrets, which are by default only remote clusters.
```

Comment:
Hey, apologies for the confusion. I made a few additional changes in #318 which improve the flow - however this should still be self-contained. I might be splitting hairs here, however I think it's more than just a reword.

> ... `getSecretsByClusterName` in your parent comment, but this function is only used to retrieve remote clusters...

This is the crux of the subtlety - `getSecretsByClusterName()` retrieves *any* cluster with a secret, which may include the local cluster. At the moment *local* implies no secret, and *remote* implies secret; the latter is always true, but the former is not, so the docs sometimes qualify this with *default local*, but not universally. Unrelated to the docs, the code comments (e.g. `If there is a secret for this cluster, then it's a non-local cluster`) suggest it was written with the belief that the local cluster would never have a secret.

Picking through what I removed:

> The cluster generator will automatically target both local and non-local clusters, for every cluster that matches the cluster selector.

It's easy for readers to misinterpret this as "any set of clusters can be matched by the selector". In reality, the selector can match both local and remote clusters, but not always at the same time. If the selector is non-empty, only clusters with a secret, which by default *excludes* the local cluster, can be targeted.

At this stage it's only a *lack* of selector that can include the local cluster, as we haven't yet said it's possible for the local cluster to have a secret. I think it's important to distinguish between a label selector not matching the local cluster, and it not actually being possible for a secret-less local cluster to be involved.

> This selector will not match the default local cluster, since the default local cluster does not have a Secret (and thus does not have the `argocd.argoproj.io/secret-type` label on that secret).

I removed this as it does not apply if a secret has been minted for the local cluster. The next section, which I've kept, mentions this.

> Any cluster selector that selects on that label will automatically exclude the default local cluster.

Any cluster selector that selects on *any* label will automatically exclude the default local cluster. This is also implied by the next section stating there is a way to include the local cluster.

To be honest, having gone through it again, I think this PR makes the docs more correct than before, but no more helpful. I'd like to make further improvements, e.g. moving the bit about creating a secret for the local cluster to a dedicated heading - it's only relevant to operators of Argo. I'd also explicitly encourage operators who use this generator to deploy both locally and remotely to create a secret to normalise the selection of clusters. Let's put this on hold for this release; I'm sure you're very busy finalising more important pieces.

Reply:
:sunglasses: , and I didn't intend to/don't want to diminish the significance of a doc reword PR: words are the fundamental atomic unit of ideas, and picking the right words to communicate the right ideas is essential to producing software that has an easy to grasp, yet powerful mental model. So I'm 100% for any tweaks that improve that, and love it when folks take their own valuable time to think through how to best to improve that!  :+1: 

I'm happy to move this to post-release, and you have definitely raised some questions in my own mind that would be beneficial to address in this generator doc page for the next release (or sooner). Among other things, I'm personally not happy with the 'local' vs 'remote' terminology, I suspect a better set of terms exists. And likewise, in the product code itself, we may want to make sure that, for example, we are properly handling the case of a local cluster with a secret (related to the code comment you referenced.)